src/storage/table/row_group_collection.cpp

```diff
@@ -836,9 +837,66 @@ void RowGroupCollection::RemoveFromIndexes(const QueryContext &context, TableInd
 		// Slice the vector with all rows that are present in this vector.
 		// If the index is bound, delete the data. If unbound, buffer into unbound_index.
 		result_chunk.Slice(sel, sel_count);
-		indexes.Scan([&](Index &index) {
+		indexes.ScanEntries([&](IndexEntry &entry) {
+			auto &index = *entry.index;
 			if (index.IsBound()) {
-				index.Cast<BoundIndex>().Delete(result_chunk, row_identifiers);
+				auto &main_index = index.Cast<BoundIndex>();
+				optional_ptr<BoundIndex> add_index, removal_index;
+				auto current_removal_type = removal_type;
+				if (!main_index.RequiresTransactionality()) {
+					// this index is not transactional - don't use "deleted_rows_in_use"
+					if (current_removal_type == IndexRemovalType::MAIN_INDEX) {
+						current_removal_type = IndexRemovalType::MAIN_INDEX_ONLY;
+					} else if (current_removal_type == IndexRemovalType::REVERT_MAIN_INDEX_APPEND) {
+						current_removal_type = IndexRemovalType::REVERT_MAIN_INDEX_ONLY_APPEND;
+					} else if (current_removal_type == IndexRemovalType::DELETED_ROWS_IN_USE) {
+						// nop
+						return false;
+					}
+				}
```

Comment:
Maybe we can put this and the below `switch` in a function or make it more clear what we're doing? I.e., all of these steps are to figure out which index to remove from, and which index to append to, if I understand correctly. Maybe we can even make that more explicit via the names of the enum? I think without this section of code its a bit hard to understand the `enum` and how its being used... 🤔